### DISCUSSION TITLE
Unable to run & build StrongSwan 5.8.0 on CentOS 7.9.2009 : public key parsing issue

### DISCUSSION THREAD

**Discussion by dhanaseelangit** [strongswan/strongswan#1640] [Q&A] [answered]
I am **compiling strongswan5.8.0 source code** for my work related to **Strongswan Road Warrior case**  based on the strongswan wiki page and hitting into the **issues related to Public key at Client side**. I have generated my key files based on strongswan PKI tool.

[root@vos2-ctl-node-002 strongswan]# cat /etc/centos-release
**CentOS Linux release 7.9.2009** (Core)

[root@vos2-ctl-node-002 strongswan]# uname -a
Linux vos2-ctl-node-002 **3.10.0-1160.el7.x86_64** #1 SMP Mon Oct 19 16:18:59 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
[root@vos2-ctl-node-002 strongswan]# 

[root@vos2-ctl-node-002 strongswan]# lsmod
Module                  Size  Used by
ipt_MASQUERADE         12678  2 
nf_nat_masquerade_ipv4    13463  1 ipt_MASQUERADE
ip6t_rpfilter          12595  1 
ip6t_REJECT            12625  2 
nf_reject_ipv6         13717  1 ip6t_REJECT
ipt_REJECT             12541  2 
nf_reject_ipv4         13373  1 ipt_REJECT
xt_conntrack           12760  21 
ebtable_nat            12807  1 
ebtable_broute         12731  1 
ip6table_nat           12864  1 
nf_conntrack_ipv6      18935  11 
nf_defrag_ipv6         35104  1 nf_conntrack_ipv6
nf_nat_ipv6            14131  1 ip6table_nat
ip6table_mangle        12700  1 
ip6table_security      12710  1 
ip6table_raw           12683  1 
iptable_nat            12875  1 
nf_conntrack_ipv4      15053  12 
nf_defrag_ipv4         12729  1 nf_conntrack_ipv4
nf_nat_ipv4            14115  1 iptable_nat
nf_nat                 26583  3 nf_nat_ipv4,nf_nat_ipv6,nf_nat_masquerade_ipv4
iptable_mangle         12695  1 
iptable_security       12705  1 
iptable_raw            12678  1 
nf_conntrack          139264  7 nf_nat,nf_nat_ipv4,nf_nat_ipv6,xt_conntrack,nf_nat_masquerade_ipv4,nf_conntrack_ipv4,nf_conntrack_ipv6
ebtable_filter         12827  1 
ebtables               35009  3 ebtable_broute,ebtable_nat,ebtable_filter
ip6table_filter        12815  1 
ip6_tables             26912  5 ip6table_filter,ip6table_mangle,ip6table_security,ip6table_nat,ip6table_raw
iptable_filter         12810  1 
ip_tables              27126  5 iptable_security,iptable_filter,iptable_mangle,iptable_nat,iptable_raw
binfmt_misc            17468  1 
veth                   13458  0 
drbg                   30186  1 
ansi_cprng             12989  0 
xfrm6_mode_tunnel      12605  0 
**xfrm4_mode_tunnel      12605  0** 
nfnetlink_queue        18197  0 
nfnetlink_log          17892  0 
bluetooth             548688  0 
rfkill                 22391  1 bluetooth
**xfrm4_tunnel           12857  0** 
**tunnel4                13252  1 xfrm4_tunnel**
ipcomp                 12700  0 
xfrm_ipcomp            13413  1 ipcomp
**esp4                   17247  0** 
**ah4                    17247  0** 
af_key                 40225  0 
br_netfilter           22256  0 
overlay                91659  0 
sr_mod                 22416  0 
cdrom                  42600  1 sr_mod
joydev                 17389  0 
cdc_ether              14682  0 
usbnet                 44378  1 cdc_ether
mii                    14434  1 usbnet
uas                    22445  0 
usb_storage            66718  1 uas
tun                    36164  0 
bridge                151336  2 br_netfilter,ebtable_broute
stp                    12976  1 bridge
llc                    14552  2 stp,bridge
ip_set                 45799  0 
nfnetlink              14519  4 nfnetlink_log,ip_set,nfnetlink_queue
bonding               152979  0 
iTCO_wdt               13572  0 
iTCO_vendor_support    13242  1 iTCO_wdt
sunrpc                358543  1 
ipmi_ssif              29595  0 
skx_edac               18261  0 
intel_powerclamp       14451  0 
coretemp               13444  0 
intel_rapl             19451  0 
iosf_mbi               15582  1 intel_rapl
kvm_intel             188740  0 
vfat                   17461  1 
fat                    65950  1 vfat
kvm                   637289  1 kvm_intel
irqbypass              13503  1 kvm
crc32_pclmul           13133  0 
ghash_clmulni_intel    13273  0 
aesni_intel           189456  0 
lrw                    13286  1 aesni_intel
gf128mul               15139  1 lrw
glue_helper            13990  1 aesni_intel
ablk_helper            13597  1 aesni_intel
cryptd                 21190  3 ghash_clmulni_intel,aesni_intel,ablk_helper
ses                    17546  0 
enclosure              15699  1 ses
scsi_transport_sas     41224  1 ses
pcspkr                 12718  0 
sg                     40719  0 
lpc_ich                21086  0 
i2c_i801               22736  0 
mei_me                 32848  0 
mei                    91150  1 mei_me
wmi                    21636  0 
ipmi_si                59965  0 
ipmi_devintf           17459  0 
ipmi_msghandler        56728  3 ipmi_ssif,ipmi_devintf,ipmi_si
tpm_crb                13639  0 
acpi_power_meter       18104  0 
acpi_pad              116316  0 
xfs                   997727  2 
libcrc32c              12644  3 xfs,nf_nat,nf_conntrack
sd_mod                 46281  4 
crc_t10dif             12912  1 sd_mod
crct10dif_generic      12647  0 
ast                    55467  1 
i2c_algo_bit           13413  1 ast
drm_kms_helper        186531  1 ast
syscopyarea            12529  1 drm_kms_helper
sysfillrect            12701  1 drm_kms_helper
sysimgblt              12640  1 drm_kms_helper
fb_sys_fops            12703  1 drm_kms_helper
ttm                    96673  1 ast
i40e                  369230  0 
ahci                   34056  0 
crct10dif_pclmul       14307  1 
crct10dif_common       12595  3 crct10dif_pclmul,crct10dif_generic,crc_t10dif
drm                   456166  4 ast,ttm,drm_kms_helper
crc32c_intel           22094  1 
libahci                31992  1 ahci
megaraid_sas          163773  3 
libata                243094  2 ahci,libahci
ptp                    23551  1 i40e
pps_core               19057  1 ptp
drm_panel_orientation_quirks    17180  1 drm
nfit                   55639  1 skx_edac
libnvdimm             159524  1 nfit
dm_mirror              22289  0 
dm_region_hash         20813  1 dm_mirror
dm_log                 18411  2 dm_region_hash,dm_mirror
dm_mod                124501  5 dm_log,dm_mirror
[root@vos2-ctl-node-002 strongswan]# 

Configure options used :
  ./configure --prefix=/usr --sysconfdir=/etc --disable-defaults --enable-charon --enable-ikev2 --enable-nonce --enable-random --enable-openssl --enable-pem --enable-x509 --enable-pubkey --enable-constraints --enable-pki --enable-socket-default --enable-kernel-netlink --enable-swanctl --enable-resolve --enable-eap-identity --enable-eap-md5 --enable-eap-dynamic --enable-eap-tls --enable-updown --enable-vici --enable-silent-rules --enable-load-tester --enable-save-keys

 strongSwan will be built with the following plugins
-----------------------------------------------------
**libstrongswan: random nonce x509 constraints pubkey pem openssl
libcharon:     load-tester kernel-netlink resolve save-keys socket-default vici updown eap-identity eap-md5 eap-dynamic eap-tls**
libtnccs:     
libtpmtss:    

Did “make all” and “make install” just followed the steps mentioned in https://github.com/strongswan/strongswan/blob/master/HACKING.

Created the Keys with Strongswan PKI tool,

 strongswan pki --gen --type ed25519 --outform pem > carolKey.pem 
strongswan pki --req --type priv --in carolKey.pem --dn "C=CH, O=strongswan, CN=carol@strongswan.org" --san carol@strongswan.org --outform pem > carolReq.pem
 strongswan pki --issue --cacert strongswanCert.pem --cakey strongswanKey.pem --type pkcs10 --in carolReq.pem --serial 01 --lifetime 1826 --outform pem > carolCert.pem

Filed copied to location as per wiki, carolKey.pem to private, carolCert.pem to X509 and CA strongswanCert.pem file to X509ca.

Swanctl.conf file content:
connections {
    home {
      local_addrs = 172.34.25.7
      remote_addrs = 172.34.25.3 
      local {
          auth = pubkey
          certs = carolCert.pem
          id = carol.strongswan.org 
        }
      remote {
        auth = pubkey
        id = moon.strongswan.org 
      }
      children {
        home {
           #local_ts  = dynamic
           #remote_ts = dynamic
           #local_ts = 10.5.0.1/32
           remote_ts = 10.1.0.1/32
           #updown = /usr/local/libexec/ipsec/_updown iptables
           esp_proposals = aes128gcm128-x25519
           #start_action = start
        }
        version = 2
        proposals = aes128-sha256-x25519
        #proposals = aes128-sha256-modp2048
      }
    }
}

[root@vos2-ctl-node-002 strongswan-5.8.0]# ./charon &
[1] 92791
[root@vos2-ctl-node-002 strongswan-5.8.0]# 00[DMN] Starting IKE charon daemon (strongSwan 5.8.0, Linux 3.10.0-1160.el7.x86_64, x86_64)
00[DMN] !!
00[DMN] !! WARNING: SAVING IKE AND ESP KEYS TO '/root/ca/keys'
00[DMN] !!
00[LIB] openssl FIPS mode(0) - disabled 
**00[LIB] loaded plugins: charon random save-keys nonce x509 constraints pubkey pem openssl kernel-netlink resolve socket-default vici updown eap-identity eap-md5 eap-dynamic eap-tls**
00[JOB] spawning 16 worker threads

[root@vos2-ctl-node-002 strongswan-5.8.0]# 

[root@vos2-ctl-node-002 strongswan-5.8.0]# swanctl --load-all
no files found matching '/etc/swanctl/conf.d/*.conf'
**04[LIB] unsupported public key algorithm
04[LIB] OpenSSL X.509 parsing failed
04[LIB] building CRED_CERTIFICATE - X509 failed, tried 4 builders
loading '/etc/swanctl/x509/carolCert.pem' failed: parsing X509 certificate failed
09[LIB] unsupported public key algorithm
09[LIB] OpenSSL X.509 parsing failed
09[LIB] building CRED_CERTIFICATE - X509 failed, tried 4 builders
loading '/etc/swanctl/x509ca/strongswanCert.pem' failed: parsing X509 certificate failed
building CRED_PRIVATE_KEY - ANY failed, tried 3 builders
13[LIB] building CRED_PRIVATE_KEY - ANY failed, tried 3 builders
loading '/etc/swanctl/private/carolKey.pem' failed: parsing ANY private key failed
no authorities found, 0 unloaded
no pools found, 0 unloaded
04[LIB] unsupported public key algorithm
04[LIB] OpenSSL X.509 parsing failed
04[LIB] building CRED_CERTIFICATE - X509 failed, tried 4 builders**
loading connection 'home' failed: invalid value for: certs, config discarded
loaded 0 of 1 connections, 1 failed to load, 0 unloaded
[root@vos2-ctl-node-002 strongswan-5.8.0]# 

The **same steps I followed in Ubuntu based Docker environment and able to make a successful IPSec tunnel and send traffic**.

As I am new to this area, I could not crack much on this. I tried checking on the online thread but I could not able to resolve the issue. Can someone Please help here?

**Comment by tobiasbrunner** [comment]
> I am **compiling strongswan5.8.0 source code**

Why such an outdated version?

> strongswan pki --gen --type ed25519 --outform pem

To create or use ED25519 keys you need at least OpenSSL 1.1.1 (which CentOS 7.9 might not ship). Or you enable the _curve25519_ plugin (run `make clean` first if you do that).

**Reply by dhanaseelangit** [reply]
Thanks for the suggestion Tobias Brunner.

I confirmed that only openssl 1.0.2 is installed as part CentOS 7.9.
 
I tried enabling the curve25519 option in configure, As suggested did make clean before building. But still seeing the same issue. 

Do you have any other suggestion. Please kindly share. I will parallelly try with latest strongswan version.

[root@vos1-ctl-node-003 strongswan-5.8.0]# ./charon 
00[DMN] Starting IKE charon daemon (strongSwan 5.8.0, Linux 3.10.0-1160.el7.x86_64, x86_64)
00[DMN] !!
00[DMN] !! WARNING: SAVING IKE AND ESP KEYS TO '/root/ca/keys'
00[DMN] !!
00[LIB] openssl FIPS mode(0) - disabled 
00[LIB] loaded plugins: charon random save-keys nonce x509 constraints pubkey pem openssl **curve25519** kernel-netlink resolve socket-default vici updown eap-identity eap-md5 eap-dynamic eap-tls
00[JOB] spawning 16 worker threads
04[LIB] unsupported public key algorithm
04[LIB] OpenSSL X.509 parsing failed
04[LIB] building CRED_CERTIFICATE - X509 failed, tried 4 builders
09[LIB] unsupported public key algorithm
09[LIB] OpenSSL X.509 parsing failed
09[LIB] building CRED_CERTIFICATE - X509 failed, tried 4 builders
14[LIB] building CRED_PRIVATE_KEY - ANY failed, tried 3 builders
15[LIB] unsupported public key algorithm
15[LIB] OpenSSL X.509 parsing failed
15[LIB] building CRED_CERTIFICATE - X509 failed, tried 4 builders

**Reply by tobiasbrunner** [reply, marked as answer]
Did you generate new keys and certificates? Also, you obviously don't have to use Ed25519 at all. You can also generate RSA or ECDSA keys/certificates.

**Reply by dhanaseelangit** [reply]
Thanks a lot Tobias Brunner. I generated a fresh RSA key based Key/Certificates and its works perfectly..
